### ISSUE TITLE
Don't change locale from C locale

### ISSUE THREAD

**Issue by Flamefire** [Return-To-The-Roots/s25client#1069]
Locale handling in C++/C seems to be "broken by design" and unreliable at best. (see https://www.boost.org/doc/libs/1_48_0/libs/locale/doc/html/std_locales.html) Issues like #429 #777 #1067 keep creeping up. It also requires "workarounds" like `helpers::toStringClassic` which use a class imbued stringstream for conversions where locale dependency is not wanted.

I therefore suggest to not change the locale from the default locale which is "C" and hence locale independent.

This requires a change in `mygettext` to reset the locale after having it set (which is required to e.g. parse the system locale) at https://github.com/Return-To-The-Roots/mygettext/blob/master/src/gettext.cpp#L76. It would however be better to NOT pretend to be a drop-in for `gettext` and significantly reduce the interface to the absolute minimum:
- set_lang
- gettext with macro `_`
- get_lang

Alternatives:   
E.g. widelands does (seem to) use standard `gettext` but has quite some code to handle changing locale. E.g. on linux they play with `LC_ALL, LANG, LANGUAGE` variables, try combinations of languages with suffixes (.utf-8 ...), calls `setlocale(LC_MESSAGES` with that and finally calls `setlocale(LC_ALL, "")` (which will use the env vars set before).   
No idea how they get away with formatting in cfg files though. No `classic` found anywhere... 

What do you think? @Flow86

**Comment by Flow86** [comment]
I know there are some problems with those locales (and yes they're broken as f... in c++)

>I therefore suggest to not change the locale from the default locale which is "C" and hence locale independent.

what about number formatting, etc? Perhaps you need it for translations for the user? For example the german translation of "0.34%" should be "0,34%".

But of course, every internal stuff (like the video settings in #1067) should not be translated at all. Why does the stringstream automagically uses the locale? thats what I think is broken there.

The main problem I think is that we do not have a separated enough Views to only translate and use the locale stuff there?

**Comment by Flamefire** [comment]
> what about number formatting, etc? Perhaps you need it for translations for the user? For example the german translation of "0.34%" should be "0,34%".

The problem here is that e.g. "2000 items" might be "translated" to "2.000 items" in German but "2 000 items" in Russian (where the space would be a non-breaking-space and "items" obviously in Russian) with the separator a possibly broken UTF-8 char (see #1067 although I don't know why)

We could do 2 things: Use C-Locale by default and use user-Locale for user facing text or user-Locale by default and C-Locale for cfgs (currently). The first is probably saver but maybe more verbose and complex especially as the latter will probably error more visibly.

Well ok, it seems we should keep it as-is :-/

BTW: Why are we using an own gettext implementation instead of the standard one?

**Comment by Flow86** [comment]
> We could do 2 things: Use C-Locale by default and use user-Locale for user facing text

yep that would be the prefered variant. but will cause a lot of work

> BTW: Why are we using an own gettext implementation instead of the standard one?

because the default gettext has some ugly specialities in catalog and locale naming between platforms. also most macros bugged like we tried to use them. Perhaps its better know, but as we started the project, it was much easier and portable this way.

**Comment by Flamefire** [comment]
Alright. Closing then as "to much work, let's fix the current code"

**Comment by Flamefire** [comment]
Opening again. So basically the problems are caused by setting `std::locale::global` which influences all stream ops, string conversions etc. This also gives us `0,13%` for German locales, which is good. But IMO this is rarely needed. Where are we using floats to display such values to user? I think we (mostly) use ints.

A problem I have found is that `LocaleHelper::createUtf8Locale` does not do that. For linux it simply uses  the system locale. Almost always this is an UTF-8 locale, but it could be e.g. `uk_UA.koi8u`. This does have a thousand separator of `0x9A` which is not an UTF-8 char (obviously as it isn't an UTF-8 locale), see https://vstinner.readthedocs.io/unicode.html

We could query `std::locale.name()`, parse this (implemented in mygettext`) check if it is an UTF-8 locale and load an UTF-8 locale instead. This fails, if the user didn't install an UTF-8 locale, but then we can show a message and fall back to C-locale.

This still opens  problems like #1067 caused by buggy locales (I assume).

Furthermore when changing the language, we change the C-API locale (`setlocale`) but not the C++-API locale (`std::locale::global`), so we always format with the system default format. I guess this needs fixing too.

Not sure, how to proceed. I remember something about `std::locale.name()` not being 100% reliable but am not sure. I guess on windows it is always empty. Something like that...

I checked Widelands again and they never touch the C++ locale, so formatting is always done with C-locale. I haven't seen many floats showing, so they simply avoid this problem. This would certainly be the easiest solution. Only point where it would be visible that I can think of is the FPS (same in Widelands) showing as `59.5` or so, but we could just round it there and be done.

Problem when going this way: It might be quite hard to go back and re-introduce the classic casts where they are required.
